File: plugins/play-billing/android/src/com/crossbow/play-billing/CrossbowPlayBilling.java
package com.crossbow.play_billing;

import com.crossbow.library.Dictionary;
import com.crossbow.library.Crossbow;
import com.crossbow.library.plugin.CrossbowPlugin;
import com.crossbow.library.plugin.SignalInfo;
import com.crossbow.library.plugin.ExposedToCrossbow;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.collection.ArraySet;

import com.android.billingclient.api.AcknowledgePurchaseParams;
import com.android.billingclient.api.AcknowledgePurchaseResponseListener;
import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.ConsumeParams;
import com.android.billingclient.api.ConsumeResponseListener;
import com.android.billingclient.api.PriceChangeConfirmationListener;
import com.android.billingclient.api.PriceChangeFlowParams;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.PurchasesResponseListener;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.android.billingclient.api.SkuDetails;
import com.android.billingclient.api.SkuDetailsParams;
import com.android.billingclient.api.SkuDetailsResponseListener;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Set;

public class CrossbowPlayBilling extends CrossbowPlugin implements PurchasesUpdatedListener, BillingClientStateListener, PriceChangeConfirmationListener {
	private final BillingClient billingClient;
	private final HashMap<String, SkuDetails> skuDetailsCache = new HashMap<>();
	private boolean calledStartConnection;
	private String obfuscatedAccountId;
	private String obfuscatedProfileId;

	public CrossbowPlayBilling(Crossbow crossbow) {
		super(crossbow);

		billingClient = BillingClient
								.newBuilder(getActivity())
								.enablePendingPurchases()
								.setListener(this)
								.build();
		calledStartConnection = false;
		obfuscatedAccountId = "";
		obfuscatedProfileId = "";
	}

	@NonNull
	@Override
	public String getPluginName() {
		return "CrossbowPlayBilling";
	}

	@NonNull
	@Override
	public Set<SignalInfo> getPluginSignals() {
		Set<SignalInfo> signals = new ArraySet<>();

		signals.add(new SignalInfo("connected"));
		signals.add(new SignalInfo("disconnected"));
		signals.add(new SignalInfo("billing_resume"));
		signals.add(new SignalInfo("connect_error", Integer.class, String.class));
		signals.add(new SignalInfo("purchases_updated", Object[].class));
		signals.add(new SignalInfo("query_purchases_response", Object.class));
		signals.add(new SignalInfo("purchase_error", Integer.class, String.class));
		signals.add(new SignalInfo("sku_details_query_completed", Object[].class));
		signals.add(new SignalInfo("sku_details_query_error", Integer.class, String.class, String[].class));
		signals.add(new SignalInfo("price_change_acknowledged", Integer.class));
		signals.add(new SignalInfo("purchase_acknowledged", String.class));
		signals.add(new SignalInfo("purchase_acknowledgement_error", Integer.class, String.class, String.class));
		signals.add(new SignalInfo("purchase_consumed", String.class));
		signals.add(new SignalInfo("purchase_consumption_error", Integer.class, String.class, String.class));

		return signals;
	}

	@Override
	public void onBillingSetupFinished(BillingResult billingResult) {
		if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
			emitSignal("connected");
		} else {
			emitSignal("connect_error", billingResult.getResponseCode(), billingResult.getDebugMessage());
		}
	}

	@Override
	public void onBillingServiceDisconnected() {
		emitSignal("disconnected");
	}

	@Override
	public void onPurchasesUpdated(final BillingResult billingResult, @Nullable final List<Purchase> list) {
		if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK && list != null) {
			emitSignal("purchases_updated", (Object)CrossbowPlayBillingUtils.convertPurchaseListToDictionaryObjectArray(list));
		} else {
			emitSignal("purchase_error", billingResult.getResponseCode(), billingResult.getDebugMessage());
		}
	}

	@Override
	public void onPriceChangeConfirmationResult(BillingResult billingResult) {
		emitSignal("price_change_acknowledged", billingResult.getResponseCode());
	}

	@Override
	public void onMainResume() {
		if (calledStartConnection) {
			emitSignal("billing_resume");
		}
	}

	@ExposedToCrossbow
	public void startConnection() {
		calledStartConnection = true;
		billingClient.startConnection(this);
	}

	@ExposedToCrossbow
	public void endConnection() {
		billingClient.endConnection();
	}

	@ExposedToCrossbow
	public boolean isReady() {
		return this.billingClient.isReady();
	}

	@ExposedToCrossbow
	public int getConnectionState() {
		return billingClient.getConnectionState();
	}

	@ExposedToCrossbow
	public void queryPurchases(String type) {
		billingClient.queryPurchasesAsync(type, new PurchasesResponseListener() {
			@Override
			public void onQueryPurchasesResponse(BillingResult billingResult,
					List<Purchase> purchaseList) {
				Dictionary returnValue = new Dictionary();
				if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
					returnValue.put("status", 0); // OK = 0
					returnValue.put("purchases", CrossbowPlayBillingUtils.convertPurchaseListToDictionaryObjectArray(purchaseList));
				} else {
					returnValue.put("status", 1); // FAILED = 1
					returnValue.put("response_code", billingResult.getResponseCode());
					returnValue.put("debug_message", billingResult.getDebugMessage());
				}
				emitSignal("query_purchases_response", (Object)returnValue);
			}
		});
	}

	@ExposedToCrossbow
	public void querySkuDetails(final String[] list, String type) {
		List<String> skuList = Arrays.asList(list);

		SkuDetailsParams.Builder params = SkuDetailsParams.newBuilder()
												  .setSkusList(skuList)
												  .setType(type);

		billingClient.querySkuDetailsAsync(params.build(), new SkuDetailsResponseListener() {
			@Override
			public void onSkuDetailsResponse(BillingResult billingResult,
					List<SkuDetails> skuDetailsList) {
				if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
					for (SkuDetails skuDetails : skuDetailsList) {
						skuDetailsCache.put(skuDetails.getSku(), skuDetails);
					}
					emitSignal("sku_details_query_completed", (Object)CrossbowPlayBillingUtils.convertSkuDetailsListToDictionaryObjectArray(skuDetailsList));
				} else {
					emitSignal("sku_details_query_error", billingResult.getResponseCode(), billingResult.getDebugMessage(), list);
				}
			}
		});
	}

	@ExposedToCrossbow
	public void acknowledgePurchase(final String purchaseToken) {
		AcknowledgePurchaseParams acknowledgePurchaseParams =
				AcknowledgePurchaseParams.newBuilder()
						.setPurchaseToken(purchaseToken)
						.build();
		billingClient.acknowledgePurchase(acknowledgePurchaseParams, new AcknowledgePurchaseResponseListener() {
			@Override
			public void onAcknowledgePurchaseResponse(BillingResult billingResult) {
				if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
					emitSignal("purchase_acknowledged", purchaseToken);
				} else {
					emitSignal("purchase_acknowledgement_error", billingResult.getResponseCode(), billingResult.getDebugMessage(), purchaseToken);
				}
			}
		});
	}

	@ExposedToCrossbow
	public void consumePurchase(String purchaseToken) {
		ConsumeParams consumeParams = ConsumeParams.newBuilder()
											  .setPurchaseToken(purchaseToken)
											  .build();

		billingClient.consumeAsync(consumeParams, new ConsumeResponseListener() {
			@Override
			public void onConsumeResponse(BillingResult billingResult, String purchaseToken) {
				if (billingResult.getResponseCode() == BillingClient.BillingResponseCode.OK) {
					emitSignal("purchase_consumed", purchaseToken);
				} else {
					emitSignal("purchase_consumption_error", billingResult.getResponseCode(), billingResult.getDebugMessage(), purchaseToken);
				}
			}
		});
	}

	@ExposedToCrossbow
	public Dictionary confirmPriceChange(String sku) {
		if (!skuDetailsCache.containsKey(sku)) {
			Dictionary returnValue = new Dictionary();
			returnValue.put("status", 1); // FAILED = 1
			returnValue.put("response_code", null); // Null since there is no ResponseCode to return but to keep the interface (status, response_code, debug_message)
			returnValue.put("debug_message", "You must query the sku details and wait for the result before confirming a price change!");
			return returnValue;
		}

		SkuDetails skuDetails = skuDetailsCache.get(sku);

		PriceChangeFlowParams priceChangeFlowParams =
			PriceChangeFlowParams.newBuilder().setSkuDetails(skuDetails).build();

		billingClient.launchPriceChangeConfirmationFlow(getActivity(), priceChangeFlowParams, this);

		Dictionary returnValue = new Dictionary();
		returnValue.put("status", 0); // OK = 0
		return returnValue;
	}

	@ExposedToCrossbow
	public Dictionary purchase(String sku) {
		return purchaseInternal("", sku,
			BillingFlowParams.ProrationMode.UNKNOWN_SUBSCRIPTION_UPGRADE_DOWNGRADE_POLICY);
	}

	@ExposedToCrossbow
	public Dictionary updateSubscription(String oldToken, String sku, int prorationMode) {
		return purchaseInternal(oldToken, sku, prorationMode);
	}

	@ExposedToCrossbow
	private Dictionary purchaseInternal(String oldToken, String sku, int prorationMode) {
		if (!skuDetailsCache.containsKey(sku)) {
			Dictionary returnValue = new Dictionary();
			returnValue.put("status", 1); // FAILED = 1
			returnValue.put("response_code", null); // Null since there is no ResponseCode to return but to keep the interface (status, response_code, debug_message)
			returnValue.put("debug_message", "You must query the sku details and wait for the result before purchasing!");
			return returnValue;
		}

		SkuDetails skuDetails = skuDetailsCache.get(sku);
		BillingFlowParams.Builder purchaseParamsBuilder = BillingFlowParams.newBuilder();
		purchaseParamsBuilder.setSkuDetails(skuDetails);
		if (!obfuscatedAccountId.isEmpty()) {
			purchaseParamsBuilder.setObfuscatedAccountId(obfuscatedAccountId);
		}
		if (!obfuscatedProfileId.isEmpty()) {
			purchaseParamsBuilder.setObfuscatedProfileId(obfuscatedProfileId);
		}
		if (!oldToken.isEmpty() && prorationMode != BillingFlowParams.ProrationMode.UNKNOWN_SUBSCRIPTION_UPGRADE_DOWNGRADE_POLICY) {
			BillingFlowParams.SubscriptionUpdateParams updateParams =
				BillingFlowParams.SubscriptionUpdateParams.newBuilder()
					.setOldSkuPurchaseToken(oldToken)
					.setReplaceSkusProrationMode(prorationMode)
					.build();
			purchaseParamsBuilder.setSubscriptionUpdateParams(updateParams);
		}
		BillingResult result = billingClient.launchBillingFlow(getActivity(), purchaseParamsBuilder.build());

		Dictionary returnValue = new Dictionary();
		if (result.getResponseCode() == BillingClient.BillingResponseCode.OK) {
			returnValue.put("status", 0); // OK = 0
		} else {
			returnValue.put("status", 1); // FAILED = 1
			returnValue.put("response_code", result.getResponseCode());
			returnValue.put("debug_message", result.getDebugMessage());
		}

		return returnValue;
	}

	@ExposedToCrossbow
	public void setObfuscatedAccountId(String accountId) {
		obfuscatedAccountId = accountId;
	}

	@ExposedToCrossbow
	public void setObfuscatedProfileId(String profileId) {
		obfuscatedProfileId = profileId;
	}
}

File: plugins/play-billing/android/src/com/crossbow/play_billing/CrossbowPlayBilling.kt
package com.crossbow.play_billing

import androidx.collection.ArraySet
import com.android.billingclient.api.AcknowledgePurchaseParams
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingFlowParams
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.ConsumeParams
import com.android.billingclient.api.PendingPurchasesParams
import com.android.billingclient.api.ProductDetails
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.PurchasesUpdatedListener
import com.android.billingclient.api.QueryProductDetailsParams
import com.android.billingclient.api.QueryPurchasesParams
import com.crossbow.library.Crossbow
import com.crossbow.library.Dictionary
import com.crossbow.library.plugin.CrossbowPlugin
import com.crossbow.library.plugin.ExposedToCrossbow
import com.crossbow.library.plugin.SignalInfo

class CrossbowPlayBilling(crossbow: Crossbow) : CrossbowPlugin(crossbow),
    PurchasesUpdatedListener, BillingClientStateListener {
    private val billingClient: BillingClient = BillingClient.newBuilder(activity)
        .enablePendingPurchases(
            PendingPurchasesParams.newBuilder()
                .enableOneTimeProducts()
                .enablePrepaidPlans()
                .build()
        )
        .enableAutoServiceReconnection()
        .setListener(this)
        .build()
    private val productDetailsCache = HashMap<String, ProductDetails>()
    private var calledStartConnection = false
    private var obfuscatedAccountId = ""
    private var obfuscatedProfileId = ""

    override val pluginName: String
        get() = javaClass.simpleName

    override val pluginSignals: Set<SignalInfo>
        get() = ArraySet<SignalInfo>().apply {
            add(SignalInfo("connected"))
            add(SignalInfo("disconnected"))
            add(SignalInfo("billing_resume"))
            add(SignalInfo("connect_error", Int::class.java, String::class.java))
            add(SignalInfo("purchases_updated", Array<Any>::class.java))
            add(SignalInfo("query_purchases_response", Any::class.java))
            add(SignalInfo("purchase_error", Int::class.java, String::class.java))
            add(SignalInfo("product_details_query_completed", Array<Any>::class.java))
            add(SignalInfo("sku_details_query_completed", Array<Any>::class.java))
            add(
                SignalInfo(
                    "product_details_query_error",
                    Int::class.java,
                    String::class.java,
                    Array<String>::class.java,
                )
            )
            add(
                SignalInfo(
                    "sku_details_query_error",
                    Int::class.java,
                    String::class.java,
                    Array<String>::class.java,
                )
            )
            add(SignalInfo("price_change_acknowledged", Int::class.java))
            add(SignalInfo("purchase_acknowledged", String::class.java))
            add(
                SignalInfo(
                    "purchase_acknowledgement_error",
                    Int::class.java,
                    String::class.java,
                    String::class.java,
                )
            )
            add(SignalInfo("purchase_consumed", String::class.java))
            add(
                SignalInfo(
                    "purchase_consumption_error",
                    Int::class.java,
                    String::class.java,
                    String::class.java,
                )
            )
        }

    override fun onBillingSetupFinished(billingResult: BillingResult) {
        if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
            emitSignal("connected")
        } else {
            emitSignal("connect_error", billingResult.responseCode, billingResult.debugMessage)
        }
    }

    override fun onBillingServiceDisconnected() = emitSignal("disconnected")

    override fun onPurchasesUpdated(billingResult: BillingResult, purchases: List<Purchase>?) {
        if (billingResult.responseCode == BillingClient.BillingResponseCode.OK && purchases != null) {
            emitSignal(
                "purchases_updated",
                CrossbowPlayBillingUtils.convertPurchaseListToDictionaryObjectArray(purchases) as Any,
            )
        } else {
            emitSignal("purchase_error", billingResult.responseCode, billingResult.debugMessage)
        }
    }

    override fun onMainResume() {
        if (calledStartConnection) emitSignal("billing_resume")
    }

    private fun failure(message: String, responseCode: Int? = null) = Dictionary().apply {
        this["status"] = 1
        this["response_code"] = responseCode
        this["debug_message"] = message
    }

    private fun purchaseInternal(
        legacyOldToken: String,
        oldProductId: String,
        productId: String,
        offerToken: String,
        replacementMode: Int,
    ): Dictionary {
        val details = productDetailsCache[productId]
            ?: return failure("Query product details and wait for the result before purchasing.")
        val productParams = BillingFlowParams.ProductDetailsParams.newBuilder()
            .setProductDetails(details)
            .apply {
                val selectedOfferToken = offerToken.ifEmpty {
                    details.subscriptionOfferDetails?.firstOrNull()?.offerToken.orEmpty()
                }
                if (
                    selectedOfferToken.isNotEmpty() &&
                    replacementMode != BillingFlowParams.ProductDetailsParams
                        .SubscriptionProductReplacementParams.ReplacementMode.KEEP_EXISTING
                ) {
                    setOfferToken(selectedOfferToken)
                }
                if (oldProductId.isNotEmpty() && replacementMode != 0) {
                    setSubscriptionProductReplacementParams(
                        BillingFlowParams.ProductDetailsParams.SubscriptionProductReplacementParams
                            .newBuilder()
                            .setOldProductId(oldProductId)
                            .setReplacementMode(replacementMode)
                            .build()
                    )
                }
            }
            .build()
        val flowParams = BillingFlowParams.newBuilder()
            .setProductDetailsParamsList(listOf(productParams))
            .apply {
                if (obfuscatedAccountId.isNotEmpty()) setObfuscatedAccountId(obfuscatedAccountId)
                if (obfuscatedProfileId.isNotEmpty()) setObfuscatedProfileId(obfuscatedProfileId)
                if (legacyOldToken.isNotEmpty()) {
                    val updateParams = BillingFlowParams.SubscriptionUpdateParams.newBuilder()
                        .setOldPurchaseToken(legacyOldToken)
                        .apply {
                            if (oldProductId.isEmpty() && replacementMode != 0) {
                                @Suppress("DEPRECATION")
                                setSubscriptionReplacementMode(replacementMode)
                            }
                        }
                        .build()
                    setSubscriptionUpdateParams(updateParams)
                }
            }
            .build()
        val result = billingClient.launchBillingFlow(activity, flowParams)
        return if (result.responseCode == BillingClient.BillingResponseCode.OK) {
            Dictionary().apply { this["status"] = 0 }
        } else {
            failure(result.debugMessage, result.responseCode)
        }
    }

    @ExposedToCrossbow
    fun startConnection() {
        calledStartConnection = true
        billingClient.startConnection(this)
    }

    @ExposedToCrossbow
    fun endConnection() = billingClient.endConnection()

    @get:ExposedToCrossbow
    val isReady: Boolean
        get() = billingClient.isReady

    @get:ExposedToCrossbow
    val connectionState: Int
        get() = billingClient.connectionState

    @ExposedToCrossbow
    fun queryPurchases(type: String) {
        val params = QueryPurchasesParams.newBuilder().setProductType(type).build()
        billingClient.queryPurchasesAsync(params) { billingResult, purchases ->
            val response = if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                Dictionary().apply {
                    this["status"] = 0
                    this["purchases"] =
                        CrossbowPlayBillingUtils.convertPurchaseListToDictionaryObjectArray(purchases)
                }
            } else {
                failure(billingResult.debugMessage, billingResult.responseCode)
            }
            emitSignal("query_purchases_response", response as Any)
        }
    }

    @ExposedToCrossbow
    fun queryProductDetails(productIds: Array<String?>, type: String) {
        val products = productIds.filterNotNull().map { productId ->
            QueryProductDetailsParams.Product.newBuilder()
                .setProductId(productId)
                .setProductType(type)
                .build()
        }
        if (products.isEmpty()) {
            emitProductQueryError(
                BillingClient.BillingResponseCode.DEVELOPER_ERROR,
                "At least one product ID is required.",
                productIds,
            )
            return
        }
        val params = QueryProductDetailsParams.newBuilder().setProductList(products).build()
        billingClient.queryProductDetailsAsync(params) { billingResult, result ->
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                result.productDetailsList.forEach { details ->
                    productDetailsCache[details.productId] = details
                }
                val converted = CrossbowPlayBillingUtils
                    .convertProductDetailsListToDictionaryObjectArray(result.productDetailsList)
                emitSignal("product_details_query_completed", converted as Any)
                // Legacy signal retained so existing Rust applications keep working.
                emitSignal("sku_details_query_completed", converted as Any)
                if (result.unfetchedProductList.isNotEmpty()) {
                    val unfetched = result.unfetchedProductList.map { it.productId }.toTypedArray()
                    emitProductQueryError(
                        BillingClient.BillingResponseCode.ITEM_UNAVAILABLE,
                        "Some products could not be fetched.",
                        unfetched,
                    )
                }
            } else {
                emitProductQueryError(
                    billingResult.responseCode,
                    billingResult.debugMessage,
                    productIds,
                )
            }
        }
    }

    @ExposedToCrossbow
    fun querySkuDetails(productIds: Array<String?>, type: String) =
        queryProductDetails(productIds, type)

    private fun emitProductQueryError(code: Int, message: String, productIds: Array<out String?>) {
        val ids = productIds.filterNotNull().toTypedArray()
        emitSignal("product_details_query_error", code, message, ids)
        emitSignal("sku_details_query_error", code, message, ids)
    }

    @ExposedToCrossbow
    fun acknowledgePurchase(purchaseToken: String) {
        val params = AcknowledgePurchaseParams.newBuilder().setPurchaseToken(purchaseToken).build()
        billingClient.acknowledgePurchase(params) { result ->
            if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                emitSignal("purchase_acknowledged", purchaseToken)
            } else {
                emitSignal(
                    "purchase_acknowledgement_error",
                    result.responseCode,
                    result.debugMessage,
                    purchaseToken,
                )
            }
        }
    }

    @ExposedToCrossbow
    fun consumePurchase(purchaseToken: String) {
        val params = ConsumeParams.newBuilder().setPurchaseToken(purchaseToken).build()
        billingClient.consumeAsync(params) { result, token ->
            if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                emitSignal("purchase_consumed", token)
            } else {
                emitSignal(
                    "purchase_consumption_error",
                    result.responseCode,
                    result.debugMessage,
                    token,
                )
            }
        }
    }

    @Deprecated("Price change confirmation was removed from Play Billing Library 9.")
    @ExposedToCrossbow
    fun confirmPriceChange(productId: String): Dictionary = failure(
        "Price change confirmation is no longer available in Play Billing Library 9 ($productId).",
        BillingClient.BillingResponseCode.DEVELOPER_ERROR,
    )

    @ExposedToCrossbow
    fun purchase(productId: String): Dictionary = purchaseInternal("", "", productId, "", 0)

    @ExposedToCrossbow
    fun purchaseWithOffer(productId: String, offerToken: String): Dictionary =
        purchaseInternal("", "", productId, offerToken, 0)

    @Deprecated("Use replaceSubscription with the old purchase token and product ID for Billing 9.")
    @ExposedToCrossbow
    fun updateSubscription(oldToken: String, productId: String, replacementMode: Int): Dictionary =
        purchaseInternal(oldToken, "", productId, "", replacementMode)

    @Deprecated("Use replaceSubscription with the old purchase token and product ID for Billing 9.")
    @ExposedToCrossbow
    fun updateSubscriptionWithOffer(
        oldToken: String,
        productId: String,
        offerToken: String,
        replacementMode: Int,
    ): Dictionary = purchaseInternal(oldToken, "", productId, offerToken, replacementMode)

    @ExposedToCrossbow
    fun replaceSubscription(
        oldPurchaseToken: String,
        oldProductId: String,
        newProductId: String,
        offerToken: String,
        replacementMode: Int,
    ): Dictionary = purchaseInternal(
        oldPurchaseToken,
        oldProductId,
        newProductId,
        offerToken,
        replacementMode,
    )

    @ExposedToCrossbow
    fun setObfuscatedAccountId(accountId: String) {
        obfuscatedAccountId = accountId
    }

    @ExposedToCrossbow
    fun setObfuscatedProfileId(profileId: String) {
        obfuscatedProfileId = profileId
    }
}
